snuba/schemas.py

```diff
@@ -217,42 +183,149 @@ def get_time_series_query_schema_properties(default_granularity: int, default_wi
     }
 }
 
-EVENTS_QUERY_SCHEMA = {
-    **copy.deepcopy(GENERIC_QUERY_SCHEMA),
+PERFORMANCE_EXTENSION_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        # Never add FINAL to queries, enable sampling
+        'turbo': {
+            'type': 'boolean',
+            'default': False,
+        },
+        # Force queries to hit the first shard replica, ensuring the query
+        # sees data that was written before the query. This burdens the
+        # first replica, so should only be used when absolutely necessary.
+        'consistent': {
+            'type': 'boolean',
+            'default': False,
+        },
+        'debug': {
+            'type': 'boolean',
+            'default': False,
+        },
+    },
+    'additionalProperties': False,
+}
+
+PROJECT_EXTENSION_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'project': {
+            'anyOf': [
+                {'type': 'number'},
+                {
+                    'type': 'array',
+                    'items': {'type': 'number'},
+                    'minItems': 1,
+                },
+            ]
+        },
+    },
     # Need to select down to the project level for customer isolation and performance
     'required': ['project'],
+    'additionalProperties': False,
 }
 
-EVENTS_QUERY_SCHEMA['properties'].update({
-    **get_time_series_query_schema_properties(
+
+def get_time_series_extension_properties(default_granularity: int, default_window: timedelta):
+    return {
+        'type': 'object',
+        'properties': {
+            'from_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: (datetime.utcnow().replace(microsecond=0) - default_window).isoformat()
+            },
+            'to_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: datetime.utcnow().replace(microsecond=0).isoformat()
+            },
+            'granularity': {
+                'type': 'number',
+                'default': default_granularity,
+            },
+        },
+        'additionalProperties': False,
+    }
+
+
+@dataclass
+class Request:
+    query: Mapping[str, Any]
+    extensions: Mapping[str, Mapping[str, Any]]
+
+    def __post_init__(self):
+        self.body = ChainMap(self.query, *self.extensions.values())
+
+
+class RequestSchema:
+    def __init__(self, query_schema, extensions_schemas: Mapping[str, Any]):
```

Comment:
isn't query_schema supposed to be of type Mapping[str, Any] as well ?

Reply:
It'd probably be a new `Schema` type, which would be aliased to … something? Probably some kind of `TypedDict`? This type annotation would then become:

```python
def __init__(self, query_schema: Schema, extensions_schemas: Mapping[str, Schema]):
```

Reply:
Added some minimal stubs with 188cbacef363fb3bc81c2533e5663b7b8a76a1c0, we can expand the type definition later if needed.